File: Easy/141-LL-cycle.py
# Runtime: 60 ms, faster than 81.66% of Python3 online submissions for Linked List Cycle.
# Memory Usage: 17.7 MB, less than 37.27% of Python3 online submissions for Linked List Cycle.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow = head
        if not head:
            return False
        fast = head.next
        
        while fast != slow:
            if not fast or not fast.next:
                return False
            else:
                fast = fast.next.next
                slow = slow.next
        return True

File: Easy/141-LLcycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *slow = head;
        if (head == nullptr) {return false;}
        ListNode *fast = head->next;
        
        while (slow != fast) {
            if (fast == nullptr || fast->next == nullptr) {
                return false;
            } else {
                slow = slow->next;
                fast = fast->next->next;
            }
        }
        
        return true;
    }
};